Include/Beman/InplaceVector/inplace_vector.hpp

```diff
@@ -0,0 +1,755 @@
+  internal_size_type size_{0};
+
+  // [containers.sequences.inplace.vector.cons], construct/copy/destroy
+  constexpr inplace_vector_destruct_base() = default;
+
+  inplace_vector_destruct_base(
+      const inplace_vector_destruct_base
+          &other) noexcept(std::is_nothrow_copy_constructible_v<T>)
+      : elems(), size_(other.size_) {}
+
+  inplace_vector_destruct_base(
+      const inplace_vector_destruct_base
+          &&other) noexcept(std::is_nothrow_move_constructible_v<T>)
+      : elems(), size_(other.size()) {}
+
+  inplace_vector_destruct_base &
+  operator=(const inplace_vector_destruct_base &other) noexcept(
+      std::is_nothrow_copy_constructible_v<T> &&
+      std::is_nothrow_copy_assignable_v<T>) {
+    size_ = other.size_;
+  }
+
+  inplace_vector_destruct_base &
+  operator=(const inplace_vector_destruct_base &&other) noexcept(
+      std::is_nothrow_move_constructible_v<T> &&
+      std::is_nothrow_move_assignable_v<T>) {
+    size_ = other.size_;
+    other.size_ = nullptr;
+  }
+
+  constexpr inplace_vector_destruct_base(const size_type size)
+      : elems(), size_(size) {
+    if (Capacity < size) {
+      throw std::bad_alloc();
+    }
+  }
+};
+
+template <typename T, std::size_t Capacity>
+struct inplace_vector_base : public inplace_vector_destruct_base<T, Capacity> {
+  using size_type = std::size_t;
+  using reference = T &;
+  using iterator = T *;
+  using const_iterator = const T *;
+
+  // [containers.sequences.inplace.vector.cons], construct/copy/destroy
+  constexpr inplace_vector_base() = default;
+  inplace_vector_base(const inplace_vector_base &other) noexcept(
+      std::is_nothrow_copy_constructible_v<T>)
+      : inplace_vector_destruct_base<T, Capacity>(other.size) {
+    std::ranges::copy(other.begin(), other.end(), begin());
+  }
+  inplace_vector_base(inplace_vector_base &&other) noexcept(
+      Capacity == 0 || std::is_nothrow_move_constructible_v<T>)
+      : inplace_vector_destruct_base<T, Capacity>(other.size) {
+    std::ranges::copy(other.begin(), other.end(), begin());
+    std::destroy(other.begin(), other.end());
+    other.size = 0;
+  }
+  inplace_vector_base &operator=(const inplace_vector_base &other) noexcept(
+      std::is_nothrow_copy_constructible_v<T> &&
+      std::is_nothrow_copy_assignable_v<T>) {
+    const auto diff = static_cast<std::ptrdiff_t>(other.size() - size());
+    // other.size is less than size
+    if (diff < 0) {
+      const auto new_end =
+          std::ranges::copy(other.begin(), other.end(), begin());
+      // destroy unnecessary memory
+      std::destroy(new_end, end());
+    }
+    // other.size is greater than size
+    else {
+      // copy other vector into the current vector until it runs ouf of size
+      std::ranges::copy(other.begin(), other.begin() + size(), begin());
+      // copy the other half after the end of the current vector
+      std::ranges::copy(other.begin() + size(), other.end(), end());
```

Comment:
The 2 copies above need the 2 iterator overload even with ranges::copy -- so this is good.